Add batched_grad parameter to `autograd.grad`

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65566
* __->__ #65564
* #65565

- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature
- adds basic test (further testing is done when we replace the usage in vectorized jacobian computation)

TODO:
 - create an issue tracking this

Differential Revision: [D31236259](https://our.internmc.facebook.com/intern/diff/D31236259)

soulitzer committed:
Add batched_grad parameter to `autograd.grad`

[ghstack-poisoned]

vadimkantorov commented on torch/autograd/__init__.py:
```diff
@@ -147,13 +150,13 @@ def backward(
     if retain_graph is None:
         retain_graph = create_graph
 
     Variable._execution_engine.run_backward(
```
Still think that the best would be to just collapse these three lines into a single one for fixing https://github.com/pytorch/pytorch/issues/41659 (even if this needs suppressing the syntax linter for this particualr case)

soulitzer replied:
Hmm, I'm okay with this

albanD commented on torch/autograd/__init__.py:
```diff
@@ -21,16 +21,19 @@
         if isinstance(grad, torch.Tensor):
-            if not out.shape == grad.shape:
+            grad_shape = grad.shape if not batched_grads else grad.shape[1:]
+            if not out.shape == grad_shape:
```
Should the error message be adapted below?

albanD commented on torch/autograd/__init__.py:
```diff
@@ -147,13 +150,13 @@ def backward(
 
     Variable._execution_engine.run_backward(
         tensors, grad_tensors_, retain_graph, create_graph, inputs,
-        allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
```
nit if this comment is for the error message, maybe mention that this is going to c++?

soulitzer replied:
Ended up going with Vadim's suggestion, but also added a comment to point this out

albanD replied:
I actually think that the comment is better than the one liner.
Pointing the users to `Variable._execution_engine` is going to lead them nowhere.

soulitzer replied:
The comment does seem to be more useful, but I guess for the more experienced user, it is more about aesthetics than usefulness lol.

If we only care about newer versions of python, the multi-line issue correctly displays the first line on Python 3.8 it seems https://bugs.python.org/issue12458.

I'm okay with just doing:
```
  File "/home/local/pytorch/torch/autograd/__init__.py", line 267, in grad
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/home/local/pytorch/torch/autograd/function.py", line 199, in apply
    return user_fn(self, *args)
  File "/home/local/test.py", line 10, in backward
    assert False
AssertionError
```

soulitzer replied:
From offline discussion: we should duplicate the comment on both lines, and add comment above

albanD commented on torch/autograd/__init__.py:
```diff
@@ -203,6 +207,18 @@ def grad(
         allow_unused (bool, optional): If ``False``, specifying inputs that were not
             used when computing outputs (and therefore their grad is always zero)
             is an error. Defaults to ``False``.
+        batched_grads (bool, optional): If ``True``, the first dimension of each
```
For boolean args, I prefer `is_grads_batched` to make it clear from the name that it is boolean and does not contain actual gradients

albanD commented on torch/autograd/__init__.py:
```diff
@@ -203,6 +207,18 @@ def grad(
+            to the autograd engine so that this computation can be performed in a
+            single call. This should lead to performance improvements when compared
+            to manually looping and performing backward multiple times. Note that
+            due to this feature being incomplete, there may be performance
```
incomplete -> experiemental?

albanD commented on torch/autograd/__init__.py:
```diff
@@ -203,6 +207,18 @@ def grad(
+            to manually looping and performing backward multiple times. Note that
+            due to this feature being incomplete, there may be performance
+            cliffs. Please use `torch._C._debug_only_display_vmap_fallback_warnings(True)`
+            to show any performance warnings and file us issues if warnings exist
```
file us -> file on github

albanD reviewed (commented):
Looks pretty good.
test though?

soulitzer committed:
Update on "Add batched_grad parameter to `autograd.grad`"


- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature

TODO:
 - create an issue tracking this

[ghstack-poisoned]

soulitzer committed:
Update on "Add batched_grad parameter to `autograd.grad`"


- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature

TODO:
 - create an issue tracking this

[ghstack-poisoned]

albanD reviewed (approved):
LGTM only duplicating the comment on the engine call to make sure it displays for all python versions.

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236259).

soulitzer committed:
Update on "Add batched_grad parameter to `autograd.grad`"


- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature
- adds basic test (further testing is done when we replace the usage in vectorized jacobian computation)

TODO:
 - create an issue tracking this

Differential Revision: [D31236259](https://our.internmc.facebook.com/intern/diff/D31236259)

[ghstack-poisoned]

soulitzer committed:
Update on "Add batched_grad parameter to `autograd.grad`"


- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature
- adds basic test (further testing is done when we replace the usage in vectorized jacobian computation)

TODO:
 - create an issue tracking this

Differential Revision: [D31236259](https://our.internmc.facebook.com/intern/diff/D31236259)

[ghstack-poisoned]

zou3519 commented on test/test_autograd.py:
```diff
@@ -719,6 +719,30 @@ def backward(ctx, x):
+        # Detect shape mismatch
+        grad_out = torch.ones(2, 2)
+        with self.assertRaisesRegex(RuntimeError, "If `is_grads_batched=True`, we interpret the first"):
+            torch.autograd.grad(outputs=out, grad_outputs=(grad_out,), inputs=(x,), is_grads_batched=True)
```
If vmap ends up in pytorch core, would we still want to have the `is_grads_batched=True` API? Because you can do this with by using vmap:
```
vmap(torch.autograd.grad, (None, None, 0))(out, x, grad_out)
```

soulitzer replied:
One reason for having both is that some users might be more comfortable with imperative APIs over a functional one,
though I can see that its not super ideal to have two ways of doing the same thing.

zou3519 replied:
Paging @albanD for discussion. Either way I'm fine with this going in as-is (we can always revisit the messaging later...)

albanD replied:
I think this is a good question. I think it is ok to have both, especially because this one is not necessarily related to vmap. And I can see people using this one without even knowing what vmap is.

zou3519 commented on torch/autograd/__init__.py:
```diff
@@ -21,21 +21,35 @@
+            grad_shape = grad.shape if not is_grads_batched else grad.shape[1:]
+            if not out.shape == grad_shape:
+                if is_grads_batched:
+                    raise RuntimeError("If `is_grads_batched=True`, we interpret the first "
+                                       "dimension of each grad_output as the batch dimension. "
+                                       "The sizes of the remaining dimensions are expected to match "
+                                       "the shape of corresponding output, but a mismatch "
```
Maybe suggest using vmap if the user actually wanted only some of their grad_output to be batched

soulitzer committed:
Update on "Add batched_grad parameter to `autograd.grad`"


- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature
- adds basic test (further testing is done when we replace the usage in vectorized jacobian computation)

TODO:
 - create an issue tracking this

Differential Revision: [D31236259](https://our.internmc.facebook.com/intern/diff/D31236259)

[ghstack-poisoned]

soulitzer committed:
Update on "Add batched_grad parameter to `autograd.grad`"


- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature
- adds basic test (further testing is done when we replace the usage in vectorized jacobian computation)

TODO:
 - create an issue tracking this

Differential Revision: [D31236259](https://our.internmc.facebook.com/intern/diff/D31236259)

[ghstack-poisoned]

soulitzer committed:
Update on "Add batched_grad parameter to `autograd.grad`"


- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature
- adds basic test (further testing is done when we replace the usage in vectorized jacobian computation)

TODO:
 - create an issue tracking this

Differential Revision: [D31236259](https://our.internmc.facebook.com/intern/diff/D31236259)

[ghstack-poisoned]

soulitzer committed:
Update on "Add batched_grad parameter to `autograd.grad`"


- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature
- adds basic test (further testing is done when we replace the usage in vectorized jacobian computation)

TODO:
 - create an issue tracking this

Differential Revision: [D31236259](https://our.internmc.facebook.com/intern/diff/D31236259)

[ghstack-poisoned]

soulitzer committed:
Update on "Add batched_grad parameter to `autograd.grad`"


- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature
- adds basic test (further testing is done when we replace the usage in vectorized jacobian computation)

TODO:
 - create an issue tracking this

Differential Revision: [D31236259](https://our.internmc.facebook.com/intern/diff/D31236259)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236259).

soulitzer committed:
Update on "Add batched_grad parameter to `autograd.grad`"


- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature
- adds basic test (further testing is done when we replace the usage in vectorized jacobian computation)

TODO:
 - create an issue tracking this

Differential Revision: [D31236259](https://our.internmc.facebook.com/intern/diff/D31236259)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236259).

soulitzer committed:
Update on "Add batched_grad parameter to `autograd.grad`"


- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature
- adds basic test (further testing is done when we replace the usage in vectorized jacobian computation)

TODO:
 - create an issue tracking this

Differential Revision: [D31236259](https://our.internmc.facebook.com/intern/diff/D31236259)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236259).
